File: 백준/Gold/10026. 적록색약/적록색약.py
import sys
from collections import deque
input = sys.stdin.readline


def bfs(r, c, people):
    q = deque()
    q.append((r, c))

    while q:
        curR, curC = q.popleft()

        for dr, dc in d:
            nextR, nextC = curR + dr, curC + dc
            if 0 <= nextR < n and 0 <= nextC < n and not visited[nextR][nextC] and people[r][c] == people[nextR][nextC]:
                visited[nextR][nextC] = True
                q.append((nextR, nextC))


n = int(input())
visionPeople = []
blindnessPeople = [['']*n for _ in range(n)]

for _ in range(n):
    visionPeople.append(list(input()))

for i in range(n):
    for j in range(n):
        if visionPeople[i][j] == 'R' or visionPeople[i][j] == 'G':
            blindnessPeople[i][j] = 'G'
        elif visionPeople[i][j] == 'B':
            blindnessPeople[i][j] = 'B'

d = [(1, 0), (0, 1), (-1, 0), (0, -1)]
visited = [[False] * n for _ in range(n)]
visionCnt = 0
for i in range(n):
    for j in range(n):
        if not visited[i][j]:
            visionCnt += 1
            bfs(i, j, visionPeople)

blindnessCnt = 0
visited = [[False] * n for _ in range(n)]
for i in range(n):
    for j in range(n):
        if not visited[i][j]:
            blindnessCnt += 1
            bfs(i, j, blindnessPeople)

print(visionCnt, blindnessCnt)

File: 백준/Gold/10026. 적록색약/적록색약.java
import java.io.*;
import java.util.*;

public class Main {

    public static int n;
    public static char[][] visionPeople;
    public static char[][] blindnessPeople;
    public static boolean[][] visited;
    public static int[][] d = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        visionPeople = new char[n][n];
        blindnessPeople = new char[n][n];

        for (int i = 0; i < n; i++) {
            String line = br.readLine();
            for (int j = 0; j < n; j++) {
                visionPeople[i][j] = line.charAt(j);
                if (visionPeople[i][j] == 'R' || visionPeople[i][j] == 'G') {
                    blindnessPeople[i][j] = 'G';
                } else if (visionPeople[i][j] == 'B') {
                    blindnessPeople[i][j] = 'B';
                }
            }
        }

        visited = new boolean[n][n];
        int visionCnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j]) {
                    visionCnt++;
                    bfs(i, j, visionPeople);
                }
            }
        }

        visited = new boolean[n][n];
        int blindnessCnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j]) {
                    blindnessCnt++;
                    bfs(i, j, blindnessPeople);
                }
            }
        }

        System.out.println(visionCnt + " " + blindnessCnt);
    }

    public static void bfs(int r, int c, char[][] people) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{r, c});
        visited[r][c] = true;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int curR = cur[0];
            int curC = cur[1];

            for (int[] direction : d) {
                int nextR = curR + direction[0];
                int nextC = curC + direction[1];

                if (nextR >= 0 && nextR < n && nextC >= 0 && nextC < n && !visited[nextR][nextC] && people[r][c] == people[nextR][nextC]) {
                    visited[nextR][nextC] = true;
                    q.add(new int[]{nextR, nextC});
                }
            }
        }
    }
}
